README.md
# DYCornerBanner

[![CI Status](http://img.shields.io/travis/dominikbutz/DYCornerBanner.svg?style=flat)](https://travis-ci.org/dominikbutz/DYCornerBanner)
[![Version](https://img.shields.io/cocoapods/v/DYCornerBanner.svg?style=flat)](http://cocoapods.org/pods/DYCornerBanner)
[![License](https://img.shields.io/cocoapods/l/DYCornerBanner.svg?style=flat)](http://cocoapods.org/pods/DYCornerBanner)
[![Platform](https://img.shields.io/cocoapods/p/DYCornerBanner.svg?style=flat)](http://cocoapods.org/pods/DYCornerBanner)


DYCornerBanner written in **Swift 3** allows you to quickly create corner labels on top of any kind of rectangular view, e.g. UIImageView. The rotation angle of the corner label is determined by the width and height of the banner view.  

## Example

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Features

* set the corner position to bottom right, top right, top left or bottom left.
* supports banners in any rotation angle and distance from the corner - the rotation angle and the distance from the corner are determined by the size and height-width proportion of the DYCornerBanner view frame. 
* override the default setting for the banner width. 
* override the default setting for the banner text font.

## Installation

DYCornerBanner is available through [CocoaPods](https://cocoapods.org/?q=dycornerbanner). To install
it, simply add the following line to your Podfile:

```ruby
pod 'DYCornerBanner'
```

Make sure to import DYCornerBanner into your View Controller subclass:

```Swift
import DYCornerBanner
```

## Usage

Check out the following examples.

![DYCornerBanner examples](./gitResources/cornerBannerExamples.png "Corner banner examples")

### Code example: Creating a DYCornerBanner from Interface Builder

```Swift
 @IBOutlet weak var cornerBannerBottomLeft: DYCornerBanner!  // simply drop a UIView into the VC from interface builder in storyboard  and set the UIView class to DYCornerBanner
 

 override func viewDidLoad() {
   super.viewDidLoad()
   
   self.cornerBannerBottomLeft.setup(text: "Sunshine Guaranteed!", textColor: 	UIColor.black, bannerColor: UIColor.yellow, position: .bottomLeft)
   
 }

```

### Code example: Creating a DYCornerBanner in code

Alternatively, you can create a corner banner instance in code like so: 
```Swift
   let frame = CGRect(x: self.imageView.frame.size.width -  90, y: 0, width: 90, height: 90)
   let bannerInCode = DYCornerBanner(frame: frame)
   bannerInCode.setup(text: "Free Cocktails!", textColor: UIColor.red, bannerColor: UIColor.white.withAlphaComponent(0.6), position: .topRight)
  self.imageView.addSubview(bannerInCode)
```


### Customising
You can override the settings for bannerWidth and textFont as shown below.  

```Swift

self.cornerBannerTopLeft.setup(text: "Just Added!", textColor: UIColor.white, bannerColor: UIColor.grassGreen(), position: .topLeft)

self.cornerBannerTopLeft.textFont = UIFont(name: "ChalkboardSE-Bold" , size: 15.0)!
self.cornerBannerTopLeft.bannerWidth = 17.0

```
## Change log

#### [Version 1.1.1](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.1.1)
moving setup of label to draw(rect:) 

#### [Version 1.1](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.1)
banner width (=label height) set explicitly now, not any more by creating a dummy label and using its height. text font and banner width can now be overridden after or before calling setup. 

#### [Version 1.0.3](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0.3)
extended documentation

#### [Version 1.0.2](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0.2)
Using cocoapods template to re-create the project.

#### [Version 1.0](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0)
Released on 2017-04-12.
First release!

## Author

dominikbutz, dominikbutz@gmail.com

## License

DYCornerBanner is available under the MIT license. See the LICENSE file for more info.




CHANGELOG.md
# Change log

## [Version 1.0.3](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0.3)
extended documentation

## [Version 1.0.2](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0.2)
Using cocoapods template to re-create the project.

## [Version 1.0](https://github.com/DominikButz/DYCornerBanner/releases/tag/1.0)
Released on 2017-04-12.
First release!




Example/Tests/Tests.swift
// https://github.com/Quick/Quick

import Quick
import Nimble
import DYCornerBanner

class TableOfContentsSpec: QuickSpec {
    override func spec() {
        
//        let imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 400, height: 200))
//        
//        self.cornerBannerBottomRight = DYCornerBanner(frame: CGRect(x: 240, y: 110, width: 160, height: 90))
//        
//        self.imageView?.image = UIImage(named:"IMG_0002.jpg")
//        self.imageView?.addSubview(self.cornerBannerBottomRight!)
//        
        context("these will pass") {
            
            it("can do maths") {
                expect(23) == 23
            }
            
            it("can read") {
                expect("🐮") == "🐮"
            }
            
            it("will eventually pass") {
                var time = "passing"
                
                DispatchQueue.main.async {
                    time = "done"
                }
                
                waitUntil { done in
                    Thread.sleep(forTimeInterval: 0.5)
                    expect(time) == "done"
                    
                    done()
                }
            }
        }
        
//        describe("these will fail") {
//
////            it("can do maths") {
////                expect(1) == 2
////            }
//
////            it("can read") {
////                expect("number") == "string"
////            }
//
////            it("will eventually fail") {
////                expect("time").toEventually( equal("done") )
////            }
//            
//
//        }
    }
}


Example/DYCornerBanner/AppDelegate.swift
//
//  AppDelegate.swift
//  DYCornerBanner
//
//  Created by dominikbutz on 04/14/2017.
//  Copyright (c) 2017 dominikbutz. All rights reserved.
//

import UIKit

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?


    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        return true
    }

    func applicationWillResignActive(_ application: UIApplication) {
        // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
        // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
    }

    func applicationDidEnterBackground(_ application: UIApplication) {
        // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
        // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
    }

    func applicationWillEnterForeground(_ application: UIApplication) {
        // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
    }

    func applicationDidBecomeActive(_ application: UIApplication) {
        // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
    }

    func applicationWillTerminate(_ application: UIApplication) {
        // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
    }


}



DYCornerBanner/Classes/DYCornerBanner.swift
//
//  DYCornerBanner.swift
//  Flat Chat
//
//  Created by Dominik Butz on 10.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import UIKit

public class DYCornerBanner: UIView {

    /// label used as subview to represent the banner
    var bannerLabel:UILabel = UILabel()
    
    /// the width of the banner  label
    var labelLength: CGFloat = 0.0
    
    /// the banner position
    var bannerPosition: BannerPosition = .bottomRight
   
    /// text of the banner label
    public var text: String!
    
    /// color of the banner label text.
  var textColor: UIColor!
    
    
    /// background color of the banner label
    var bannerColor: UIColor!
    

    fileprivate let configuration = Configuration()
    
    /// text font, can be overridden
    open var textFont: UIFont {
        get {
            return self.configuration.textFont
        }
        set(font) {
            self.configuration.textFont = font
        }
    }
    
        /// banner width,  can be overridden
    open var bannerWidth: CGFloat {
        get {
            return self.configuration.bannerWidth
        }
        set(width) {
            self.configuration.bannerWidth = width
        }
    }

    

    
    /// This method gets called if you create the view in the Interface Builder
    required public init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
        print("init decoder")
        
        
  
    }
    
    /// This method gets called if you create the view in code
  public  override init(frame: CGRect){
        super.init(frame: frame)
    print("init in code with frame")
    

    }
    
    /// setup - Has to be called on corner banner instance after creating the instance either in code or in the storyboard
    ///
    /// - Parameters:
    ///   - text: Banner label text.
    ///   - textColor: text color.
    ///   - bannerColor: fill colour of the banner label.
    ///   - position: .bottomRight, .topRight, .topLeft or .bottomLeft
    open func setup(text: String, textColor: UIColor, bannerColor: UIColor, position:BannerPosition ) {
        
        // self
        self.backgroundColor = UIColor.clear
        self.isUserInteractionEnabled = false
        self.clipsToBounds = true
        
        // setup instance variables
        self.text = text
        self.textColor = textColor
        self.bannerColor = bannerColor
        self.bannerPosition = position
        
        
        self.labelLength =  sqrt( pow(self.frame.width, 2) +  pow(self.frame.height, 2))
        
    }
    
    
    public override func draw(_ rect: CGRect) {
        super.draw(rect)
        
        print("currrent version with draw rect")
        let labelFrame = CGRect(x: 0, y: 0, width: self.labelLength, height: self.bannerWidth)
        
        self.bannerLabel = UILabel(frame: labelFrame)
        
        
        //  banner config
        self.bannerLabel.text = self.text
        self.bannerLabel.textAlignment = .center
        self.bannerLabel.baselineAdjustment = .alignCenters
        self.bannerLabel.font = self.textFont
        self.bannerLabel.textColor = self.textColor
        self.bannerLabel.numberOfLines = 1
        self.bannerLabel.lineBreakMode = .byTruncatingTail
        self.bannerLabel.backgroundColor = self.bannerColor
        
        // banner positioning
        var xAdjustment: CGFloat
        var yAdjustment: CGFloat
        let halfLabelHeight: CGFloat  = self.bannerWidth / 2.0
        
        switch self.bannerPosition  {
        case .bottomRight:
            xAdjustment = halfLabelHeight
            yAdjustment = halfLabelHeight
        case .topRight:
            xAdjustment = halfLabelHeight
            yAdjustment = -halfLabelHeight
        case .bottomLeft:
            xAdjustment = -halfLabelHeight
            yAdjustment = halfLabelHeight
        case .topLeft:
            xAdjustment = -halfLabelHeight
            yAdjustment = -halfLabelHeight
            
        }
        
        self.bannerLabel.center = CGPoint(x: self.frame.width / 2.0 + xAdjustment, y: self.frame.height / 2.0 + yAdjustment)
        
        //print("label \(self.bannerLabel)")
        
        
        let sinRotationAngle = self.frame.size.height / self.bannerLabel.frame.size.width
        let arcRotation = asin(sinRotationAngle)
        // print("arc rotation \(arcRotation)")
        
        let rotationAngleFactor = { () -> CGFloat in
            switch self.bannerPosition {
            case .bottomRight, .topLeft:
                return -1.0
            case .bottomLeft, .topRight:
                return 1.0
            }
        }()
        
        let rotation = CGAffineTransform.init(rotationAngle: CGFloat(rotationAngleFactor * arcRotation))
        
        self.bannerLabel.transform = rotation
        
        
        self.addSubview(self.bannerLabel)

        
    }


}



/// an enum for setting the corner banner position
///
/// - bottomRight:  surprise! bottomRight corner.
/// - topRight: topRight
/// - topLeft: yes, you guessed it: top left corner!
/// - bottomLeft: wow, this one, too!
public enum BannerPosition {
    case bottomRight, topRight, topLeft, bottomLeft
}



fileprivate class Configuration {
    
    var textFont: UIFont!
    var bannerWidth: CGFloat!
    
    init() {
       self.setupDefaults()
    }
    
    /// defaults for bannerWidth (= banner label height) and text font. The properties in the corner label instance can be overridden.
    func setupDefaults() {
        
        self.bannerWidth = 18.0
        self.textFont = UIFont(name: "HelveticaNeue-Medium", size: 10.0)
        
    }
    
    
}


Example/DYCornerBanner/ViewController.swift
//
//  ViewController.swift
//  DYCornerBannerDemo
//
//  Created by Dominik Butz on 11.04.17.
//  Copyright © 2017 Duoyun. All rights reserved.
//

import UIKit
import DYCornerBanner

class ViewController: UIViewController {

    @IBOutlet weak var imageView: UIImageView!
    
    @IBOutlet weak var cornerBannerBottomRight: DYCornerBanner!
    
    @IBOutlet weak var cornerBannerTopLeft: DYCornerBanner!
    
    @IBOutlet weak var cornerBannerBottomLeft: DYCornerBanner!
    

    override func viewDidLoad() {
        super.viewDidLoad()
        
        
        // with custom Font
   
        self.cornerBannerBottomRight.setup(text: "Top Cruise Tour!", textColor: UIColor.white, bannerColor: UIColor.red, position: .bottomRight)
        self.cornerBannerBottomRight.textFont = UIFont(name: "AmericanTypewriter-Bold", size: 11.0)!
    
        
        // with custom font and custom width
        self.cornerBannerTopLeft.textFont = UIFont(name: "ChalkboardSE-Bold" , size: 15.0)!
 
        self.cornerBannerTopLeft.setup(text: "Just Added!", textColor: UIColor.white, bannerColor: UIColor.grassGreen(), position: .topLeft)
       self.cornerBannerTopLeft.bannerWidth = 25.0
       
        
        self.cornerBannerBottomLeft.setup(text: "Sunshine Guaranteed!", textColor: UIColor.black, bannerColor: UIColor.yellow, position: .bottomLeft)
        
        
        // programatically created DY Corner Banner
        let frame = CGRect(x: self.imageView.frame.size.width -  90, y: 0, width: 90, height: 90)
        let bannerInCode = DYCornerBanner(frame: frame)
       bannerInCode.textFont = UIFont(name: "HelveticaNeue-Bold", size: 10.0)!
        bannerInCode.setup(text: "Free Cocktails!", textColor: UIColor.red, bannerColor: UIColor.white.withAlphaComponent(0.6), position: .topRight)
        self.imageView.addSubview(bannerInCode)
        

        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}


public extension UIColor {
    
    class  func veryDarkGray()->UIColor {
        
        return UIColor(red: 64/255, green: 64/255, blue: 65/255, alpha: 1.0)
    }
    
    class func veryLightGray()->UIColor {
        
        return UIColor(red: 239/255, green: 239/255, blue: 244/255, alpha: 1.0)
    }
    
    
    
    class func dirtyWhite()->UIColor {
        
        return UIColor(red: 250/255, green: 250/255, blue: 250/255, alpha: 1.0)
    }
    
    class func grassGreen()->UIColor {
        return UIColor(red: 39/255, green: 161/255, blue: 60/255, alpha: 1.0)
    }
    
    
    
    
}



Example/DYCornerBanner_ExampleTests/DYCornerBanner_ExampleTests.swift
//
//  DYCornerBanner_ExampleTests.swift
//  DYCornerBanner_ExampleTests
//
//  Created by Dominik Butz on 16.04.17.
//  Copyright © 2017 CocoaPods. All rights reserved.
//

import XCTest
@testable import DYCornerBanner
import UIKit

class DYCornerBannerTests: XCTestCase {

    var imageView:UIImageView?
    var cornerBannerBottomRight : DYCornerBanner?

    override func setUp() {
        super.setUp()

        self.imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 400, height: 200))

        self.cornerBannerBottomRight = DYCornerBanner(frame: CGRect(x: 240, y: 110, width: 160, height: 90))

        self.imageView?.image = UIImage(named:"IMG_0002.jpg")
        self.imageView?.addSubview(self.cornerBannerBottomRight!)


        // Put setup code here. This method is called before the invocation of each test method in the class.
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()

        self.imageView = nil
        self.cornerBannerBottomRight = nil

    }

    func testCornerBannerBottomRight() {



        self.cornerBannerBottomRight?.textFont = UIFont(name: "AmericanTypewriter-Bold", size: 11.0)!
        self.cornerBannerBottomRight?.bannerWidth = 17.0
        self.cornerBannerBottomRight?.setup(text: "Top Cruise Tour!", textColor: UIColor.white, bannerColor: UIColor.red, position: .bottomRight)

        XCTAssert(self.cornerBannerBottomRight?.bannerPosition == .bottomRight, "corner banner is wrong")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.font.familyName, "American Typewriter", "wrong font name!")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerWidth, 17.0, "Banner width value not correct")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.text, "Top Cruise Tour!", "Banner label text incorrect!")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.textColor, UIColor.white, "Banner label text color value not correct")
        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.backgroundColor, UIColor.red, "Banner label background color value not correct")


        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }

    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }

}
//
//class DYCornerBannerSpec: QuickSpec {
//    
//    
//    var imageView:UIImageView?
//    var cornerBannerBottomRight : DYCornerBanner?
//    
//    override func spec() {
//        
//        // self.setup()
//        
//        
//        self.imageView = UIImageView(frame: CGRect(x: 0, y: 0, width: 400, height: 200))
//        
//        self.cornerBannerBottomRight = DYCornerBanner(frame: CGRect(x: 240, y: 110, width: 160, height: 90))
//        
//        self.imageView?.image = UIImage(named:"IMG_0002")
//        self.imageView?.addSubview(self.cornerBannerBottomRight!)
//        
//        
//        self.cornerBannerBottomRight?.textFont = UIFont(name: "AmericanTypewriter-Bold", size: 11.0)!
//        self.cornerBannerBottomRight?.bannerWidth = 17.0
//        self.cornerBannerBottomRight?.setup(text: "Top Cruise Tour!", textColor: UIColor.white, bannerColor: UIColor.red, position: .bottomRight)
//        
//        
//        
//        describe("testing if properties satisfiy expected values") {
//            
//            it("position should be bottom right", closure: {
//                expect(self.cornerBannerBottomRight?.bannerPosition ) == .bottomRight
//                
//            })
//            
//            
//            it("font familiy should be American Tyewriter", closure: {
//                expect(self.cornerBannerBottomRight?.bannerLabel.font.familyName ) == "American Typewriter"
//                
//            })
//            
//            it("banner width should be 17.0", closure: {
//                expect(self.cornerBannerBottomRight?.bannerWidth ) == 17.0
//                
//            })
//            
//            
//            
//            //        XCTAssert(self.cornerBannerBottomRight?.bannerPosition == .bottomRight, "corner banner is wrong")
//            //        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.font.familyName, "American Typewriter", "wrong font name!")
//            //        XCTAssertEqual(self.cornerBannerBottomRight?.bannerWidth, 17.0, "Banner width value not correct")
//            //        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.text, "Top Cruise Tour!", "Banner label text incorrect!")
//            //        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.textColor, UIColor.white, "Banner label text color value not correct")
//            //        XCTAssertEqual(self.cornerBannerBottomRight?.bannerLabel.backgroundColor, UIColor.red, "Banner label background color value not correct")
//            
//        }
//        
//    }
//    
//}
//
//
////
//class TableOfContentsSpec: QuickSpec {
//    
//    
//    override func spec() {
//        describe("these will fail") {
//            
//            it("can do maths") {
//                expect(1) == 2
//            }
//            
//            it("can read") {
//                expect("number") == "string"
//            }
//            
//            it("will eventually fail") {
//                expect("time").toEventually( equal("done") )
//            }
//            
//            context("these will pass") {
//                
//                it("can do maths") {
//                    expect(23) == 23
//                }
//                
//                it("can read") {
//                    expect("🐮") == "🐮"
//                }
//                
//                it("will eventually pass") {
//                    var time = "passing"
//                    
//                    DispatchQueue.main.async {
//                        time = "done"
//                    }
//                    
//                    waitUntil { done in
//                        Thread.sleep(forTimeInterval: 0.5)
//                        expect(time) == "done"
//                        
//                        done()
//                    }
//                }
//            }
//        }
//    }
//}
